src/snowflake/snowpark/mock/_udf.py

```diff
@@ -0,0 +1,97 @@
+#
+# Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+#
+from types import ModuleType
+from typing import Callable, Dict, List, Optional, Tuple, Union
+
+from snowflake.snowpark._internal.udf_utils import (
+    check_python_runtime_version,
+    process_registration_inputs,
+)
+from snowflake.snowpark._internal.utils import TempObjectType
+from snowflake.snowpark.exceptions import SnowparkSQLException
+from snowflake.snowpark.types import DataType
+from snowflake.snowpark.udf import UDFRegistration, UserDefinedFunction
+
+
+class MockUDFRegistration(UDFRegistration):
+    def __init__(self, *args, **kwargs) -> None:
+        super().__init__(*args, **kwargs)
+        self._registry: Dict[str, Callable] = dict()
+
+    def register_from_file(self, *_, **__) -> UserDefinedFunction:
+        raise NotImplementedError(
+            "[Local Testing] Registering UDF from file is not currently supported."
+        )
+
+    def _do_register_udf(
+        self,
+        func: Union[Callable, Tuple[str, str]],
+        return_type: Optional[DataType],
+        input_types: Optional[List[DataType]],
+        name: Optional[str],
+        stage_location: Optional[str] = None,
+        imports: Optional[List[Union[str, Tuple[str, str]]]] = None,
+        packages: Optional[List[Union[str, ModuleType]]] = None,
+        replace: bool = False,
+        if_not_exists: bool = False,
+        parallel: int = 4,
+        max_batch_size: Optional[int] = None,
+        from_pandas_udf_function: bool = False,
+        strict: bool = False,
+        secure: bool = False,
+        external_access_integrations: Optional[List[str]] = None,
+        secrets: Optional[Dict[str, str]] = None,
+        immutable: bool = False,
+        *,
+        statement_params: Optional[Dict[str, str]] = None,
+        source_code_display: bool = True,
+        api_call_source: str,
+        skip_upload_on_content_match: bool = False,
+        is_permanent: bool = False,
+    ) -> UserDefinedFunction:
+        if is_permanent:
+            raise NotImplementedError(
```

Comment:
Let's use `MockConnection.log_not_supported_error` for Local Testing telemetry.

Reply:
good catch, let's replace `NotImplementedError` with `log_not_supported_error` in other places as well